### DISCUSSION TITLE
RLS vulnerability - `getUser()` does not remove the JWT when the associated session is invalid

### DISCUSSION THREAD

**Discussion by park-jemin** [supabase/supabase#26851] [Questions]
I discovered recently that from the `@supabase/ssr`, while `getUser` is recommended to validate the token, it doesn't actually remove the token when the user session no longer exists in the db. While it correctly sets the cookie when it has to refresh (and deletes it if the refresh token doesn't exist), it doesn't do so when the session itself is no longer valid. 

This results in queries backed by RLS to _not_ be protected via the recommended auth flows, as a malicious actor could highjack a valid JWT while not having a valid session. This also makes it so that invalidating sessions on other devices via `signOut()` do not purge the cookie for devices that didn't call `signOut()`, resulting in potentially unexpected behavior.

An example flow from a project I'm working on:
- User signs onto two different browsers, A and B (two sessions, two different JWTs) 
- User has access to a protected resource backed by RLS
- User signs out of browser A with global scope (invalidating all sessions)
- User no longer has access to the protected resource in browser A (and does not have the auth token stored in cookies), but they have access to the protected resource in browser B
  - In browser B, while `getUser` will return `null` for the user, the auth cookie will still persist

Is this by design? I couldn't find a lot of discussion on this specifically, so guidance here would be appreciated!

My workaround for this via Next.js middleware is to do this (slightly modified from the [guide here](https://supabase.com/docs/guides/auth/server-side/nextjs)):

```ts
// utils/supabase/middleware.ts
import { createServerClient, type CookieOptions } from '@supabase/ssr'
import { NextResponse, type NextRequest } from 'next/server'

export async function updateSession(request: NextRequest) {
  let response = NextResponse.next({
    request: {
      headers: request.headers,
    },
  })

  const supabase = createServerClient(
    process.env.NEXT_PUBLIC_SUPABASE_URL!,
    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
    {
      cookies: { 
          ...
      },
    }
  )

  const {
    data: { user },
  } = await supabase.auth.getUser()

  // NOTE: `supabase.auth.getUser()` currently does NOT remove the auth cookie if invalid. We need to do this manually.
  if (user === null) {
    // Where AUTH_TOKEN_KEY = the associated cookie name (by default sb-<project_id>-auth-token)
    request.cookies.set({ name: AUTH_TOKEN_KEY, value: '' }) 

    response = NextResponse.next({
      request: {
        headers: request.headers,
      },
    })

    response.cookies.set({ name: AUTH_TOKEN_KEY, value: '', maxAge: 0 })
  }

  return response
}
```

**Comment by GaryAustin1** [comment]
I'll just throw out this is probably more an auth-helpers/SSR thing.  Supabase-js with getSession/getUser are "generic" functions used for browsers, servers etc.   There is no knowledge of cookies for user sessions in supabase-js.  

In general the JWT should be assumed to be valid until it expires if it is ever exposed publicly (like on a browser) as it can't be cancelled.
